### PR TITLE
[risk=low][RW-6328] Remove angular dependencies & convert to react-scripts server

### PR DESCRIPTION
**Pull request by als364**
Issues I ran into when switching to react-scripts:
- react-scripts runs on Babel, which needs `isolatedModules=true` in tsconfig in order to work. [This causes problems with a number of things](https://www.typescriptlang.org/tsconfig#isolatedModules), including:
  - const enums (e.g. ts-key-enum)
  - re-exporting interfaces
- Changed all spread children to some other thing
- Changed module imports where we only used one class in the module to class imports (e.g. `Color`, `validate()`, `moment()`. [ref](https://stackoverflow.com/questions/35706164/typescript-import-as-vs-import-require/35706271#35706271))
- Added some type params to some useStates and other parameterized things that we never put type params on
- Moved [stylesheets](https://create-react-app.dev/docs/adding-a-stylesheet/), [images, and fonts](https://create-react-app.dev/docs/adding-images-fonts-and-files/) to imports
- Updated e2e to handle the changes in how we import images (the src will become something like `static/media/filename-bunchanumbers.jpg`, so we instead use data-test-id)
- Many jest/webpack related changes
  - Moved jest config into config-overrides.js
  - Jest transformer to change image file imports into image file names

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/app/pages/workspace/workspace-edit.spec.tsx [hunk 1]
```diff
@@ -37,7 +37,11 @@ import SpyInstance = jest.SpyInstance;
 
 type AnyWrapper = (ShallowWrapper|ReactWrapper);
 
-jest.mock('app/utils/workbench-gapi-client', () => ({
+jest.mock('app/utils/navigation', () => ({
```
Comment (als364) [inline ui/src/app/pages/workspace/workspace-edit.spec.tsx:40]: I can't figure out how to make the import mock and the regular mock work at the same time. Please advise :(

File: ui/src/index.ts [hunk 2]
```diff
@@ -9,6 +9,22 @@ import {ErrorHandler} from 'app/components/error-handler';
 import {setupCustomValidators} from 'app/services/setup';
 import {AppRoutingComponent} from './app/routing/app-routing';
 
+import "../node_modules/@clr/icons/clr-icons.min.css";
```
Comment (ericsong) [inline ui/src/index.ts:12]: interesting, why do we have to specify the `node_modules` in the path to import the packages here?

File: ui/public/index.html [hunk 3]
```diff
@@ -25,7 +25,7 @@
   <body id='body'>
     <noscript>
       <div style="max-width: 500px; margin: 1rem;">
-        <div><img alt="logo" src="/assets/images/all-of-us-logo.svg" width="155px"></div>
+        <div><img alt="logo" src="/src/assets/images/all-of-us-logo.svg" width="155px"></div>
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/public/index.html:28]: I'd have to either eject from create-react-app or pull in react-app-rewired (which I think I want to do anyways because I think it will make it easier to handle having 8 different environments)

File: ui/src/app/cohort-search/search-group-item/search-group-item.component.tsx [hunk 4]
```diff
@@ -141,7 +140,7 @@ interface State {
   loading: boolean;
   paramListOpen: boolean;
   renaming: boolean;
-  timeout: Timeout;
+  timeout: NodeJS.Timeout;
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/cohort-search/search-group-item/search-group-item.component.tsx:143]: It was, yes. Typescript compiler couldn't figure out the type otherwise. I think that either Angular may have had a Timeout built in, or previously IntelliJ was looking at `window.setTimeout` or something. The error I was getting was `Type 'number' is not assignable to type 'Timeout'`.

File: ui/src/app/cohort-search/tree/tree.component.tsx [hunk 5]
```diff
@@ -363,7 +365,7 @@ export const CriteriaTree = fp.flow(withCurrentWorkspace(), withCurrentConcept()
             Ingredients in this brand: {ingredients.join(', ')}
           </div>}
           <button style={styles.returnLink} onClick={() => back()}>
-            <img src='/assets/icons/arrow-left-regular.svg' style={styles.arrowIcon} alt='Go back' />
+            <img src={arrow} style={styles.arrowIcon} alt='Go back' />
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/cohort-search/tree/tree.component.tsx:368]: Arrow is a variable representing the actual directory from which the assets are served, which ends up being something like `/static/media/arrow-left-regular-bunchanumbers.svg`

File: ui/src/app/components/buttons.tsx [hunk 6]
```diff
@@ -8,7 +8,7 @@ import {reactStyles} from 'app/utils/index';
 import {useNavigation} from 'app/utils/navigation';
 import * as fp from 'lodash/fp';
 import * as React from 'react';
-import * as Interactive from 'react-interactive';
+import Interactive from 'react-interactive';
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/components/buttons.tsx:11]: The jest tests break with the following error:
```Error: Uncaught [Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.
      
      Check the render method of `Clickable`.]```

Has to do with how webpack handles imports: https://stackoverflow.com/questions/34130539/uncaught-error-invariant-violation-element-type-is-invalid-expected-a-string

File: ui/src/app/components/resource-actions-menu.tsx [hunk 7]
```diff
@@ -12,7 +12,7 @@ interface Action {
   hoverText?: string;
 }
 
-const ResourceActionsMenu = (props: { actions: Action[] }) => {
+export const ResourceActionsMenu = (props: { actions: Action[] }) => {
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/components/resource-actions-menu.tsx:15]: Yes, this is due to Babel requiring `isolatedModules=true` to be set in tsconfig

File: ui/src/app/pages/admin/admin-workspace.tsx [hunk 8]
```diff
@@ -197,13 +197,13 @@ const FileDetailsTable = (props: FileDetailsProps) => {
         style={styles.fileDetailsTable}
         scrollable={true}
         paginator={true}
-        paginatorTemplate='CurrentPageReport FirstPageLink PrevPageLink PageLinks NextPageLink LastPageLink RowsPerPageDropdown'
+        paginatorTemplate='Link PrevPageLink PageLinks NextPageLink LastPageLink RowsPerPageDropdown'
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/pages/admin/admin-workspace.tsx:200]: ope this is probably a bad merge conflict resolution by me

File: ui/src/app/pages/admin/admin-workspace.tsx [hunk 9]
```diff
@@ -68,16 +68,16 @@ const styles = reactStyles({
...
-const PurpleLabel = ({style = {}, children}) => {
+const PurpleLabel = ({style = {}, labelText}) => {
   return <label style={{color: colors.primary, ...style}}>
-    {...children}
+    {labelText}
   </label>;
 };
 
-const WorkspaceInfoField = ({labelText, children}) => {
+const WorkspaceInfoField = ({labelText, content}) => {
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/pages/admin/admin-workspace.tsx:77]: Got a red squiggly about spread children but I can probably figure out a way to make it work, actually

File: ui/src/app/pages/data/cohort-review/annotation-definition-modals.component.tsx [hunk 10]
```diff
@@ -14,7 +14,6 @@ import {reactStyles, summarizeErrors} from 'app/utils';
 import { WorkspaceData } from 'app/utils/workspace-data';
 import {AnnotationType, CohortAnnotationDefinition} from 'generated/fetch';
 import { matchPath, RouteComponentProps, withRouter } from 'react-router-dom';
-import {Key} from 'ts-key-enum';
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/pages/data/cohort-review/annotation-definition-modals.component.tsx:17]: No, but `isolatedModules=true` breaks imported enums

File: ui/src/app/pages/data/data-component.tsx [hunk 11]
```diff
@@ -177,7 +176,7 @@ export const DataComponent = withCurrentWorkspace()((props: Props) => {
             {/*Because the container can stretch based on window size, but the height
               can't we set a max width to cap the height based on aspect ratio*/}
             <div style={{width: '100%', maxWidth: '425px', paddingTop: '1rem'}}>
-              <img src={cohortImg}/>
+              <img data-test-id={"cohort-diagram"} src={cohortImg}/>
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/pages/data/data-component.tsx:179]: I'll see if I can compare the image directly.

File: ui/src/index.ts (hunk 2 above)
Comment (als364) [reply ui/src/index.ts:12]: Likely there's something wrong with my configuration.

File: ui/public/index.html [hunk 12]
```diff
@@ -5,7 +5,7 @@
     <base href="/">
     <meta charset="UTF-8">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="icon" type="image/x-icon" href="favicon.ico">
+    <link rel="icon" type="image/x-icon" href="src/favicon.ico">
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/public/index.html:8]: doesn't seem to matter.

**Review by calbach — COMMENTED** [review]
FYI: I pulled this branch locally. The hot reloading is really awesome. Errors show up very cleanly as well in browser - which is a big QOL improvement. Compiled asset size looks reasonable.

File: ui/src/app/components/resource-card.tsx [hunk 13]
```diff
@@ -117,7 +118,7 @@ const ResourceNavigation = (props: NavProps) => {
 
   return <Clickable disabled={!canNavigate()}>
     <RouterLink to={url} style={style} data-test-id={linkTestId} onClick={() => onNavigate()}>
-      {...children}
+      {children}
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/components/resource-card.tsx:121]: Yeah, I was getting a compilation error `Spread children are not supported in react`.  [long list of feature requests](https://github.com/facebook/jsx/issues/57)

Dunno why it was working before.

File: ui/config-overrides.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
+const webpack = require('webpack');
+
+module.exports = {
+  webpack: function(config, env) {
+    let environmentFilePath = 'environment.localtest.ts'
+    if (process.env.REACT_APP_ENVIRONMENT === 'local' ) {
+      environmentFilePath = 'environment.local.ts'
```
Comment (ericsong) [inline ui/config-overrides.js:7]: nit: we can try something like `environmentFilePath = 'environment.{process.env.REACT_APP_ENVIRONMENT}.ts'`

File: README.md [hunk 15]
```diff
@@ -161,14 +161,14 @@ yarn dev-up
 
 You can view your local UI server at http://localhost:4200/.  
 
-By default, this connects to our test API server. Use `--configuration=$ENV` to
+By default, this connects to our test API server. Prepend `REACT_APP_ENVIRONMENT=$ENV` to
```
Comment (ericsong) [inline README.md:164]: nit: I would reword to something like.. "Prepend..." -> "Set the `REACT_APP_ENVIRONMENT` environment variable"

**Review by ericsong — APPROVED** [review]
LGTM! Nice job on the final push
